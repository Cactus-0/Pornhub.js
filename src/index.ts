import type { RequestInit } from 'node-fetch'
import type { AlbumSearchOptions, AutoCompleteOptions, GifSearchOptions, PornstarSearchOptions, VideoSearchOptions } from './types'
import { Route, getMainPage, login, logout } from './apis'
import { Engine } from './core/engine'
import { WebMaster } from './core/webmaster'
import { albumPage } from './scrapers/pages/album'
import { photoPage } from './scrapers/pages/photo'
import { videoPage } from './scrapers/pages/video'
import { albumSearch } from './scrapers/search/album'
import { pornstarSearch } from './scrapers/search/pornstar'
import { gifSearch } from './scrapers/search/gif'
import { videoSearch } from './scrapers/search/video'
import { modelSearch } from './scrapers/search/model'
import { getAutoComplete } from './apis/autoComplete'
import { getToken } from './apis/getToken'
import { modelPage } from './scrapers/pages/model'
import { pornstarPage } from './scrapers/pages/pornstar'

export class PornHub {
    engine = new Engine()
    route = Route
    webMaster = new WebMaster(this.engine)

    setAgent(agent: RequestInit['agent']) {
        this.engine.request.setAgent(agent)
    }

    setHeader(key: string, value: string) {
        this.engine.request.setHeader(key, value)
    }

    setCookie(key: string, value: any) {
        this.engine.request.setCookie(key, value)
    }

    /**
     * I don't know why this works, but it does.
     * See: https://github.com/pionxzh/Pornhub.js/issues/27
     */
    warmup() {
        return getMainPage(this.engine)
    }

    /**
     * Login with account and password.
    */
    login(account: string, password: string) {
        return login(this.engine, account, password)
    }

    /**
     * Logout from Pornhub.com.
    */
    logout() {
        return logout(this.engine)
    }

    /**
     * Get token from Pornhub.com.
     * Most of pornhub's api need this token.
     * You can cache this token to avoid frequent requests (I'm not sure about the expiration time!).
     *
     * For now, this token is only used for `autoComplete` and `searchModel`.
     * This library will automatically get token if you don't provide it.
     */
    getToken() {
        return getToken(this.engine)
    }

    /**
     * Get video information by url/ID
     * @param urlOrId Video ID or page url
    */
    video(urlOrId: string) {
        return videoPage(this.engine, urlOrId)
    }

    /**
     * Get album information by url/ID
     * @param urlOrId Album ID or page url
    */
    album(urlOrId: string) {
        return albumPage(this.engine, urlOrId)
    }

    /**
     * Get photo information by url/ID
     * @param urlOrId Photo ID or page url
    */
    photo(urlOrId: string) {
        return photoPage(this.engine, urlOrId)
    }

    /**
     * Get pornstar information by url/ID
     * @param urlOrName Pornstar name or page url
    */
    pornstar(urlOrName: string) {
        return pornstarPage(this.engine, urlOrName)
    }

    /**
     * Get model information by url/ID
     * @param urlOrName Model name or page url
     */
    model(urlOrName: string) {
        return modelPage(this.engine, urlOrName)
    }

    /**
     * Get autocomplete result by keyword.
     */
    autoComplete(keyword: string, options: AutoCompleteOptions = {}) {
        return getAutoComplete(this.engine, keyword, options)
    }

    /**
     * Search album by keyword.
     */
    searchAlbum(keyword: string, options: AlbumSearchOptions = {}) {
        return albumSearch(this.engine, keyword, options)
    }

    /**
     * Search gif by keyword.
     */
    searchGif(keyword: string, options: GifSearchOptions = {}) {
        return gifSearch(this.engine, keyword, options)
    }

    /**
     * Search pornstar by keyword.
     */
    searchPornstar(keyword: string, options: PornstarSearchOptions = {}) {
        return pornstarSearch(this.engine, keyword, options)
    }

    /**
     * Search model by keyword.
     */
    searchModel(keyword: string, options: AutoCompleteOptions = {}) {
        return modelSearch(this.engine, keyword, options)
    }

    /**
     * Search video by keyword.
     */
    searchVideo(keyword: string, options: VideoSearchOptions = {}) {
        return videoSearch(this.engine, keyword, options)
    }
}

