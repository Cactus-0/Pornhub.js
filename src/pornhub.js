const Engine = require('./core/engine')
const WebMaster = require('./webmaster')

const VideoSearch = require('./entity/search/video')
const AlbumSearch = require('./entity/search/album')
const GifSearch = require('./entity/search/gif')
const PornstarSearch = require('./entity/search/pornstar')

class PornHub {
    constructor () {
        this.engine = new Engine()
        this.engine.start()

        this.webMaster = new WebMaster(this.engine)
    }

    on (event, fn) {
        this.engine.on(event, fn)
    }

    setAgent (data) {
        this.engine.request.setAgent(data)
    }

    setHeader (data) {
        this.engine.request.setHeader(data)
    }

    login (account, password) {
        return this.engine.action.login(account, password)
    }

    logout () {
        return this.engine.action.logout()
    }

    get (url) {
        const VideoInfo = require('./entity/info/video')
        return new VideoInfo(this.engine, url).run()
    }

    video (url) {
        const VideoInfo = require('./entity/info/video')
        return new VideoInfo(this.engine, url).run()
    }

    album (url) {
        const AlbumInfo = require('./entity/info/album')
        return new AlbumInfo(this.engine, url).run()
    }

    photo (url) {
        const PhotoInfo = require('./entity/info/photo')
        return new PhotoInfo(this.engine, url).run()
    }

    /**
     * Search video by keyword.
     * @url https://www.pornhub.com/webmaster/search?search=keyword
     * @param {('Video'|'Album'|'Gif'|'Pornstar')} type Search type
     * @param {string} keyword
     * @param {Object} options
     * @param {number} [options.page=1]
     * @param {boolean} [options.hd] show hd video
     * @param {boolean} [options.verified] show verified video
     * @param {('gay'|'transgender')} [options.sexualOrientation]
     * @param {'production'|'homemade'} [options.production]
     * @param {number} [options.durationMin] Options: 10|20|30
     * @param {number} [options.durationMax] Options: 10|20|30
     * @param {('Most Relevant'|'Most Recent'|'Most Viewed'|'Top Rated'|'Longest'|'No. of Video')} [options.order]
     * @param {Array<('male'|'female'|'straight'|'gay'|'transgender'|'uncategorized'|'misc')>} [options.segments]
     * @returns {Promise.<Object[]>}
    */
    search (type, keyword, options) {
        switch (type.toLowerCase()) {
            case 'video':
                return new VideoSearch(this.engine, keyword, options).run()
            case 'gif':
                return new GifSearch(this.engine, keyword, options).run()
            case 'album':
                return new AlbumSearch(this.engine, keyword, options).run()
            case 'pornstar':
                return new PornstarSearch(this.engine, keyword, options).run()
            default:
                console.error(`Search type ${type} is not supported.`)
        }
    }
}

module.exports = PornHub
