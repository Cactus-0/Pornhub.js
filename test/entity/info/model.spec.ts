import { describe, expect, it } from 'vitest'
import { PornHub } from '../../../src/index'

const pornhub = new PornHub()

describe('Model Info', () => {
    it('# run()', async () => {
        const url = 'https://www.pornhub.com/model/luna-okko'
        const result = await pornhub.model(url)

        const {
            rank,
            subscribers,
            videoViews,
            profileViews,
            videoWatched,
            ...restResult
        } = result

        expect(result.name).toBe('Luna Okko')
        expect(rank).not.toBe(0)
        expect(subscribers).not.toBe(0)
        expect(videoViews).not.toBe(0)
        expect(profileViews).not.toBe(0)
        expect(videoWatched).not.toBe(0)

        expect(restResult).toMatchSnapshot()
    })
})
