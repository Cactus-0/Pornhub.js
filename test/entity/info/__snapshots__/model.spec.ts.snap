// Vitest Snapshot v1

exports[`Model Info > # run() 1`] = `
{
  "about": "I travel the world with my boyfriend, bringing you with us in our adventures!",
  "avatar": "https://ei.phncdn.com/pics/users/u/001/117/083/341/avatar1617043303/(m=ewILGCjadOf)(mh=mXhAG2kUEGMLb_iT)200x200.jpg",
  "bio": "",
  "ethnicity": "Asian",
  "fakeBoobs": false,
  "featuredIn": [],
  "gender": "Female",
  "hairColor": "Black",
  "height": "5' 2\\" (157cm)",
  "interestedIn": "Guys and Girls",
  "name": "Luna Okko",
  "piercings": false,
  "relationship": "Taken",
  "socials": {
    "amazonWishList": undefined,
    "instagram": "https://www.instagram.com/lunaokko",
    "modelhub": undefined,
    "snapchat": undefined,
    "twitter": "https://www.twitter.com/lunaokko_",
    "website": "https://lunaokko.com",
  },
  "tattoos": true,
  "verified": true,
  "weight": "88lbs. (40kg)",
}
`;
